### PR TITLE
 membuffer: support staging & checkpoint for ART

### PR DESCRIPTION
**Pull request by you06**
ref https://github.com/pingcap/tidb/issues/55287

This PR implements `Staging/Release/Cleanup` and `Checkpoint/RevertToCheckpoint` for ART, the behavior is exactly same with RBT.

### CODE REVIEW

File: internal/unionstore/art/art.go [hunk 1]
```diff
@@ -392,51 +393,86 @@ func (t *ART) Size() int {
 }
 
 func (t *ART) checkpoint() arena.MemDBCheckpoint {
-	panic("unimplemented")
+	return t.allocator.vlogAllocator.Checkpoint()
 }
 
 func (t *ART) RevertNode(hdr *arena.MemdbVlogHdr) {
-	panic("unimplemented")
+	lf := t.allocator.getLeaf(hdr.NodeAddr)
```
Comment (cfzjywxk) [inline internal/unionstore/art/art.go:400]: Better to check the validness before using it, so the possbile wrong state would not pass around.
```
if lf == nil
   panic("To revert invalid node")
```

**Review by cfzjywxk — APPROVED** [review]

File: internal/unionstore/art/art.go [hunk 2]
```diff
@@ -392,51 +393,89 @@ func (t *ART) Size() int {
...
+	if h > len(t.stages) {
+		return
+	}
+	if h < len(t.stages) {
+		panic(fmt.Sprintf("cannot cleanup staging buffer, h=%v, len(db.stages)=%v", h, len(t.stages)))
+	}
 
-func (t *ART) truncate(snap *arena.MemDBCheckpoint) {
-	panic("unimplemented")
+	cp := &t.stages[h-1]
```
Comment (ekexium) [inline internal/unionstore/art/art.go:469]: Is it possible that `len(stages)=0` and `h=0`? Shall we test that?

File: internal/unionstore/art/art.go [hunk 3]
```diff
@@ -392,51 +393,89 @@ func (t *ART) Size() int {
...
-
-func (t *ART) revertToCheckpoint(cp *arena.MemDBCheckpoint) {
-	panic("unimplemented")
-}
-
-func (t *ART) moveBackCursor(cursor *arena.MemDBCheckpoint, hdr *arena.MemdbVlogHdr) {
-	panic("unimplemented")
-}
+	if h > len(t.stages) {
+		return
```
Comment (ekexium) [inline internal/unionstore/art/art.go:463]: I wonder whether there are such usages...

File: internal/unionstore/art/art.go [hunk 4]
```diff
@@ -392,51 +393,89 @@ func (t *ART) Size() int {
 }
 
 func (t *ART) checkpoint() arena.MemDBCheckpoint {
-	panic("unimplemented")
+	return t.allocator.vlogAllocator.Checkpoint()
 }
 
 func (t *ART) RevertNode(hdr *arena.MemdbVlogHdr) {
```
Comment (ekexium) [inline internal/unionstore/art/art.go:399]: Is "RevertVLogEntry" or something similar a more appropriate name? `RevertNode` looks like we are deleting the node.

**Review by ekexium — COMMENTED** [review]
Can we have unit tests that test multi-level stages? Current tests only includes 1 levels and simple staging or cleanup.
The rest LGTM.

File: internal/unionstore/art/art.go (hunk 2 above)
Comment (you06) [reply internal/unionstore/art/art.go:469]: `h == 0` means staging is not support in [TiDB](https://github.com/pingcap/tidb/blob/adbc0034ffb4470b42dc9f75ea595fae7499008b/pkg/table/tables/mutation_checker.go#L88-L90), I think we can just return for this case.

I'll add test for it.

File: internal/unionstore/art/art.go (hunk 3 above)
Comment (you06) [reply internal/unionstore/art/art.go:463]: TiDB always call [`defer memBuffer.Clean`](https://github.com/pingcap/tidb/blob/fb9b8a194f8716beaee035cd0c285c30e29a9d64/pkg/table/tables/tables.go#L798) after a group of mutations (even `memBuffer.Release` is already invoked), so this case is expected and valid.

File: internal/unionstore/art/art.go (hunk 4 above)
Comment (you06) [reply internal/unionstore/art/art.go]: Rename it to `RevertVAddr`.

**Review by ekexium — APPROVED** [review]

**Pull request merged**